raiden-ts/src/services/epics.ts

```diff
@@ -903,3 +749,297 @@ export const msMonitorNewBPEpic = (
+      iou: undefined,
+    });
+  } else if (pfsIsDisabled(action, latest)) {
+    throw new RaidenError(ErrorCodes.PFS_DISABLED);
+  } else {
+    return getRouteFromPfs$(action, deps, latest);
+  }
+}
+
+function validateRoute$(
+  action: pathFind.request,
+  deps: RaidenEpicDeps,
+  route: Route,
+  latest: Latest,
+): Observable<pathFind.success | pathFind.failure | iouPersist | iouClear> {
+  const { tokenNetwork } = action.meta;
+  const { iou, paths, error } = route;
+
+  return from(
+    // looks like mergeMap with generator doesn't handle exceptions correctly
+    // use from+iterator from iife generator instead
+    (function* () {
+      if (iou) {
+        if (shouldPersistIou(route)) {
+          yield iouPersist({ iou }, { tokenNetwork: tokenNetwork, serviceAddress: iou.receiver });
+        } else {
+          yield iouClear(undefined, {
+            tokenNetwork: tokenNetwork,
+            serviceAddress: iou.receiver,
+          });
+        }
+      }
+
+      if (error) {
+        if (isNoRouteFoundError(error)) {
+          throw new RaidenError(ErrorCodes.PFS_NO_ROUTES_BETWEEN_NODES);
+        } else {
+          throw new RaidenError(ErrorCodes.PFS_ERROR_RESPONSE, {
+            errorCode: error.error_code,
+            errors: error.errors,
+          });
+        }
+      }
+
+      const filteredPaths = filterPaths(action, deps, latest, paths);
+
+      if (filteredPaths.length) {
+        yield pathFind.success({ paths: filteredPaths }, action.meta);
+      } else {
+        throw new RaidenError(ErrorCodes.PFS_NO_ROUTES_FOUND);
+      }
+    })(),
+  );
+}
+
+function validateRouteTargetAndEventuallyThrow(action: pathFind.request, latest: Latest): void {
+  const { tokenNetwork, target } = action.meta;
+  const { state, presences } = latest;
+
+  if (!Object.values(state.tokens).includes(tokenNetwork))
+    throw new RaidenError(ErrorCodes.PFS_UNKNOWN_TOKEN_NETWORK, { tokenNetwork });
+
+  if (!presences[target].payload.available)
+    throw new RaidenError(ErrorCodes.PFS_TARGET_OFFLINE, { target });
+
+  if (presences[target].payload.caps?.[Capabilities.NO_RECEIVE])
+    throw new RaidenError(ErrorCodes.PFS_TARGET_NO_RECEIVE, { target });
+}
+
+function pfsIsDisabled(action: pathFind.request, latest: Latest): boolean {
+  const disabledByAction = action.payload.pfs === null;
+  const disabledByConfig = !action.payload.pfs && latest.config.pfs === null;
+  return disabledByAction || disabledByConfig;
+}
+
+function getRouteFromPfs$(
+  action: pathFind.request,
+  deps: RaidenEpicDeps,
+  latest: Latest,
+): Observable<Route> {
+  const { tokenNetwork, target, value } = action.meta;
+
+  return getPsfInfo$(action.payload.pfs, latest.config.pfs, deps).pipe(
+    mergeMap((pfs) => getIouForPfs(pfs, tokenNetwork, deps)),
+    mergeMap(({ pfs, iou }) =>
+      requestPfs$(
+        pfs,
+        iou,
+        tokenNetwork,
+        deps.address,
+        target,
+        UInt(32).encode(value),
```

Comment:
Nitpick: can you pass the actual `value`, and encode it inside `requestPfs$`, like the others, just so we get it typeguarded there as well?

Reply:
Actually I wanted to do so first. My issue was/is that `value` is of type `t.Branded<BigNumber, UIntB<32>>`. This is a bit awkward to define as function parameter. And since `value` is retrieved from the `meta` object of a `pathFind.request`, I have no chance to use that for the function signature. At least I always get the the error `Property 'meta' does not exist on type 'ActionCreator<"path/find/request"...` But probably my knowledge is just too limited here. Could you help me?

Reply:
There would be a few ways to extract the type directly ton the action creator or type interface, some of them could require you to use `typeof`, but in this case, it's not needed, `t.Branded<BigNumber, UIntB<32>>` is actually `UInt<32>`, and they're inter-assignable. Actually, since #1870 , the type of `value` should effectively be `UInt<32>`, you can use it in the param of the function.

Reply:
:heavy_check_mark: